File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/modules/CallModule.kt
package com.rajutechie.streamkit.modules

import com.rajutechie.streamkit.transport.HttpClient
import com.rajutechie.streamkit.transport.WebSocketClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow

data class Call(val id: String, val type: CallType, val status: String, val initiatedBy: String)
data class CallConfig(val type: CallType, val participants: List<String>, val channelId: String? = null)
enum class CallType { AUDIO, VIDEO }

class CallModule(
    private val http: HttpClient,
    private val ws: WebSocketClient,
    private val scope: CoroutineScope
) {
    suspend fun start(config: CallConfig): Call = http.post("/calls", config)
    suspend fun get(callId: String): Call = http.get("/calls/$callId")
    suspend fun accept(callId: String) { http.post<Unit>("/calls/$callId/accept") }
    suspend fun reject(callId: String, reason: String? = null) { http.post<Unit>("/calls/$callId/reject", mapOf("reason" to reason)) }
    suspend fun end(callId: String) { http.post<Unit>("/calls/$callId/end") }
    suspend fun startRecording(callId: String) { http.post<Unit>("/calls/$callId/recording/start") }
    suspend fun stopRecording(callId: String) { http.post<Unit>("/calls/$callId/recording/stop") }

    fun toggleAudio(callId: String, enabled: Boolean) {
        ws.emit("call.signal", mapOf("callId" to callId, "action" to "toggle_audio", "enabled" to enabled))
    }

    fun toggleVideo(callId: String, enabled: Boolean) {
        ws.emit("call.signal", mapOf("callId" to callId, "action" to "toggle_video", "enabled" to enabled))
    }

    fun switchCamera(callId: String) {
        ws.emit("call.signal", mapOf("callId" to callId, "action" to "switch_camera"))
    }

    suspend fun startScreenShare(callId: String) {
        http.post<Unit>("/calls/$callId/screen-share/start")
    }

    suspend fun stopScreenShare(callId: String) {
        http.post<Unit>("/calls/$callId/screen-share/stop")
    }

    suspend fun getStats(callId: String): CallStats = http.get("/calls/$callId/stats")

    fun observeIncomingCalls(): Flow<Call> = ws.on("call.incoming", Call::class.java)

    fun observeCallEvents(): Flow<CallEvent> = ws.on("call.event", CallEvent::class.java)
}

data class CallStats(
    val callId: String,
    val audioPacketsLost: Int,
    val videoPacketsLost: Int,
    val audioJitter: Double,
    val videoJitter: Double,
    val roundTripTime: Double,
    val bitrate: Double,
)

data class CallEvent(val callId: String, val type: String, val data: Map<String, Any>?)

File: sdks/ios/Sources/RajutechieStreamKit/Modules/CallModule.swift
import Foundation

public struct Call: Codable {
    public let id: String
    public let type: String
    public let status: String
    public let initiatedBy: String
}

public struct CallConfig: Encodable {
    public let type: String
    public let participants: [String]
    public let channelId: String?

    public init(type: String = "video", participants: [String], channelId: String? = nil) {
        self.type = type
        self.participants = participants
        self.channelId = channelId
    }
}

public final class CallModule {
    private let http: HTTPClient
    private let ws: WebSocketClient

    init(http: HTTPClient, ws: WebSocketClient) {
        self.http = http
        self.ws = ws
    }

    public func start(_ config: CallConfig) async throws -> Call {
        let result: Call = try await http.post("/calls", body: config)
        return result
    }

    public func get(_ callId: String) async throws -> Call {
        let result: Call = try await http.get("/calls/\(callId)")
        return result
    }

    public func accept(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/accept")
    }

    public func reject(_ callId: String, reason: String? = nil) async throws {
        struct RejectCallBody: Encodable {
            let reason: String
        }
        let _: EmptyResponse = try await http.post("/calls/\(callId)/reject", body: RejectCallBody(reason: reason ?? ""))
    }

    public func end(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/end")
    }

    public func startRecording(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/recording/start")
    }

    public func stopRecording(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/recording/stop")
    }

    public func toggleAudio(_ callId: String, enabled: Bool) {
        Task { await ws.send(type: "call.signal", data: ["callId": callId, "action": "toggle_audio", "enabled": enabled]) }
    }

    public func toggleVideo(_ callId: String, enabled: Bool) {
        Task { await ws.send(type: "call.signal", data: ["callId": callId, "action": "toggle_video", "enabled": enabled]) }
    }

    public func switchCamera(_ callId: String) {
        Task { await ws.send(type: "call.signal", data: ["callId": callId, "action": "switch_camera"]) }
    }

    public func startScreenShare(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/screen-share/start")
    }

    public func stopScreenShare(_ callId: String) async throws {
        let _: EmptyResponse = try await http.post("/calls/\(callId)/screen-share/stop")
    }

    public func getStats(_ callId: String) async throws -> CallStats {
        let result: CallStats = try await http.get("/calls/\(callId)/stats")
        return result
    }

    /// Async stream of incoming calls. Yields each new `Call` as it arrives.
    public var incomingCalls: AsyncStream<Call> {
        AsyncStream { continuation in
            Task {
                await ws.on("call.incoming") { data in
                    if let call = try? JSONDecoder().decode(Call.self, from: data) {
                        continuation.yield(call)
                    }
                }
            }
        }
    }
}

public struct CallStats: Codable {
    public let callId: String
    public let audioPacketsLost: Int
    public let videoPacketsLost: Int
    public let audioJitter: Double
    public let videoJitter: Double
    public let roundTripTime: Double
    public let bitrate: Double
}
